containers/source-build/source_build.py

```diff
@@ -0,0 +1,455 @@
+    )
+    parser.add_argument(
+        "-s",
+        "--source-dir",
+        required=True,
+        dest="source_dir",
+        metavar="PATH",
+        type=arg_type_path,
+        help="Path to the directory holding source code from which to build the binary image.",
+    )
+    parser.add_argument(
+        "--output-binary-image",
+        required=True,
+        metavar="IMAGE",
+        help="The output binary image used to generate source image.",
+    )
+    parser.add_argument(
+        "--base-images",
+        metavar="IMAGES",
+        type=arg_type_base_images,
+        default="",
+        help="Base images used to build the binary image, from which to get the sources. "
+        "Each image per line and only the last (bottom) one is handled, which is the "
+        "single image or the last image specified in a multistage Dockerfile. If omitted, "
+        "skip handling sources of base image.",
+    )
+    parser.add_argument(
+        "--cachi2-artifacts-dir",
+        metavar="PATH",
+        help="Path to Cachi2 directory which is the output directory populated by fetch-deps "
+        "command and the generated environment file.",
+    )
+    parser.add_argument(
+        "--write-result-to",
+        metavar="FILE",
+        dest="result_file",
+        help="Write execution result into this file.",
+    )
+    return parser.parse_args()
+
+
+def registry_has_image(image: str) -> bool:
+    cmd = ["skopeo", "inspect", "--raw", f"docker://{image}"]
+    return run(cmd, capture_output=True).returncode == 0
+
+
+def create_dir(*components) -> str:
+    path = os.path.join(*components)
+    os.makedirs(path)
+    return path
+
+
+def extract_blob_member(
+    tar_archive: str, member: str, dest_dir: str, rename_to: str, work_dir: str, log: logging.Logger
+) -> None:
+    """Extract a blob member and rename it."""
+    # strip 3 components: ./blobs/sha256
+    tar_cmd = [
+        "tar",
+        "--extract",
+        "-C",
+        dest_dir,
+        "--strip-components",
+        "3",
+        "-f",
+        tar_archive,
+        member,
+    ]
+    log.debug("extract blob member %r", tar_cmd)
+    run(tar_cmd, check=True, cwd=work_dir)
+    shutil.move(f"{dest_dir}/{os.path.basename(member)}", f"{dest_dir}/{rename_to}")
+
+
+def prepare_base_image_sources(
+    image: str, work_dir: str, sib_dirs: SourceImageBuildDirectories
+) -> bool:
+    log = logging.getLogger("source-build.base-image-sources")
+    parts = urllib.parse.urlparse(f"docker://{image}")
+    if parts.netloc != "registry.access.redhat.com":
+        log.info(
+            "Image %s does not come from supported registry. Skip handling the sources from it.",
+            image,
+        )
+        return False
+
+    base_image_sources_dir = create_dir(work_dir, "base_image_sources")
+    base_sources_extraction_dir = create_dir(base_image_sources_dir, "extraction_dir")
+
+    source_image_name = image + "-source"
```

Comment:
I think this is not quite that simple. For example, for image = `registry.access.redhat.com/ubi9/ubi:latest`, there is no `registry.access.redhat.com/ubi9/ubi:latest-source`

```bash
$ skopeo inspect --raw docker://registry.access.redhat.com/ubi9/ubi:latest-source
FATA[0000] Error parsing image name "docker://registry.access.redhat.com/ubi9/ubi:latest-source": reading manifest latest-source in registry.access.redhat.com/ubi9/ubi: manifest unknown
```

But you can find the source image like this:

```bash
$ skopeo inspect --config docker://registry.access.redhat.com/ubi9/ubi:latest | jq '.config.Labels | "\(.version)-\(.release)"' -r
9.2-755.1697625012

$ skopeo inspect --raw docker://registry.access.redhat.com/ubi9/ubi:9.2-755.1697625012-source
...
```

That is: find the unique `$version-$release` tag, append `-source` to it -> that's the source image (I believe that's how Conjur tags them)

Reply:
Created a story for this. Thanks.